Library/Homebrew/cmd/link.rb

```diff
@@ -43,26 +43,22 @@ def link
     }
 
     kegs = if args.HEAD?
-      args.named.to_kegs.group_by(&:name).transform_values do |v|
-        v.find do |keg|
-          keg.version.head?
+      args.named.to_kegs.group_by(&:name).map do |name, resolved_kegs|
+        head_keg = resolved_kegs.find { |keg| keg.version.head? }
+        if head_keg.blank?
+          opoo <<~EOS
+            No HEAD keg installed for #{name}
+            To install, run:
+              brew install --HEAD #{name}
+          EOS
         end
-      end
+        head_keg
+      end.reject(&:blank?)
```

Comment:
This will do what you want more elegantly (and `typecheck` should be okay with it)

```suggestion
      end.compact
```

Other than this I don't think it's particularly inelegant. Was that the part you weren't a fan of?

Reply:
Unsure why, just didn't feel super clean to me. I like the second suggestion you gave better.

Why does `typecheck` realize that `compact` removes nil elements but not `reject(&:blank?)`? Is it because the latter relies on a `proc`?

Reply:
No idea honestly 🤷

If I had to guess it's because `compact` is a built in method defined for arrays. `blank?` is not built in but is pulled in by the `activesupport` gem (at least I think)

Reply:
Yeah, I don't know enough about how sorbet works to say. My guess is that it might not be able to resolve type guards in `procs` passed as `blocks`, but unsure.